lib/rules/no-unassigned-vars.js

```diff
@@ -0,0 +1,64 @@
+/**
+ * @fileoverview Rule to flag variables that are never assigned
+ * @author Jacob Bandes-Storch <https://github.com/jtbandes>
+ */
+"use strict";
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+/** @type {import('../types').Rule.RuleModule} */
+module.exports = {
+	meta: {
+		type: "problem",
+		dialects: ["typescript", "javascript"],
+		language: "javascript",
+
+		docs: {
+			description:
+				"Disallow variable declarations that use `let` or `var` but have no initial value and are never assigned",
+			recommended: false,
+			url: "https://eslint.org/docs/latest/rules/no-unassigned-vars",
+		},
+
+		schema: [],
+		messages: {
+			unassigned:
+				"'{{name}}' is not initialized and never assigned, so it will always be undefined.",
+		},
+	},
+
+	create(context) {
+		const sourceCode = context.sourceCode;
+
+		return {
+			[`VariableDeclaration:matches([kind=let], [kind=var]):not([declare=true]) > VariableDeclarator[id.type=Identifier]:not([init])`](
```

Comment:
This is a pretty long selector, which also means it will be fairly slow. Can we change this to just `VariableDeclarator` and then write the conditions in JavaScript? Something like this?

```js
const declaration = sourceCode.getParent(node);
const shouldCheck = !node.init && node.id.type === "Identifier" && declaration.kind !== "const" && !declaration.declare;
```